Source/WebCore/accessibility/AccessibilityObject.cpp

```diff
@@ -3578,37 +3584,39 @@ static int computeBestScrollOffset(int currentScrollOffset, int subfocusMin, int
 }
 
 bool AccessibilityObject::isOnScreen() const
-{   
-    bool isOnscreen = true;
-
+{
     // To figure out if the element is onscreen, we start by building of a stack starting with the
     // element, and then include every scrollable parent in the hierarchy.
-    Vector<const AccessibilityObject*> objects;
-    
+    Vector<RefPtr<const AccessibilityObject>> objects;
+
     objects.append(this);
-    for (AccessibilityObject* parentObject = this->parentObject(); parentObject; parentObject = parentObject->parentObject()) {
-        if (parentObject->getScrollableAreaIfScrollable())
-            objects.append(parentObject);
+    for (RefPtr ancestor = parentObject(); ancestor; ancestor = ancestor->parentObject()) {
+        if (ancestor->getScrollableAreaIfScrollable())
+            objects.append(WTFMove(ancestor));
     }
 
     // Now, go back through that chain and make sure each inner object is within the
     // visible bounds of the outer object.
     size_t levels = objects.size() - 1;
-    
     for (size_t i = levels; i >= 1; i--) {
         RefPtr outer = objects[i];
         RefPtr inner = objects[i - 1];
         // FIXME: unclear if we need LegacyIOSDocumentVisibleRect.
         const IntRect outerRect = i < levels ? snappedIntRect(outer->boundingBoxRect()) : outer->getScrollableAreaIfScrollable()->visibleContentRect(ScrollableArea::LegacyIOSDocumentVisibleRect);
-        const IntRect innerRect = snappedIntRect(inner->isScrollView() ? inner->parentObject()->boundingBoxRect() : inner->boundingBoxRect());
-        
-        if (!outerRect.intersects(innerRect)) {
-            isOnscreen = false;
-            break;
+
+        IntRect innerRect;
+        if (inner->isScrollView())
+            innerRect = snappedIntRect(inner->parentObject()->boundingBoxRect());
+        else {
+            innerRect = snappedIntRect(inner->boundingBoxRect());
+            if (RefPtr scrollView = !outer->isRoot() ? outer->scrollView() : nullptr)
+                innerRect = scrollView->contentsToRootView(innerRect);
```

Comment:
Can you do this without the "if"? Can contentsToRootView just be a no-op for the root view?

Reply:
> Can you do this without the "if"? Can contentsToRootView just be a no-op for the root view?

I played around with this idea, and I think the answer is no unfortunately. Despite the method name implying no work should be done if this scroll view is the root, it will unconditionally call `ScrollView::contentsToView` on the passed in rect, which results in the wrong behavior (failing layout tests, broken Voice Control in actual Safari).